atomic_reactor/plugins/build_orchestrate_build.py

```diff
@@ -0,0 +1,189 @@
+
+    def get_cluster_info(self, cluster, platform):
+        osbs = OSBS(Configuration(conf_section=cluster.name),
+                    Configuration(conf_section=cluster.name))
+        current_builds = self.get_current_builds(osbs)
+        load = current_builds / cluster.max_concurrent_builds
+        return ClusterInfo(cluster, platform, osbs, load)
+
+    def choose_cluster(self, platform):
+        config = get_config(self.workflow)
+        clusters = [self.get_cluster_info(cluster, platform) for cluster in
+                    config.get_enabled_clusters_for_platform(platform)]
+
+        if not clusters:
+            raise RuntimeError('No clusters found for platform {}!'
+                               .format(platform))
+
+        selected = min(clusters, key=lambda c: c.load)
+        self.log.info('platform %s will use cluster %s',
+                      platform, selected.cluster.name)
+        return selected
+
+    def get_release(self):
+        labels = df_parser(self.workflow.builder.df_path, workflow=self.workflow).labels
+        return get_preferred_label(labels, 'release')
+
+    def get_worker_build_kwargs(self, release, platform):
+        build_kwargs = deepcopy(self.build_kwargs)
+
+        build_kwargs.pop('architecture', None)
+
+        build_kwargs['release'] = release
+        build_kwargs['platform'] = platform
+
+        return build_kwargs
+
+    def run(self):
+        release = self.get_release()
+        worker_builds = []
+
+        clusters = [self.choose_cluster(platform)
+                    for platform in self.get_platforms()]
+
+        for cluster_info in clusters:
+            kwargs = self.get_worker_build_kwargs(release, cluster_info.platform)
+            worker_build = cluster_info.osbs.create_worker_build(**kwargs)
+            worker_builds.append(
+                WorkerBuildInfo(build=worker_build, cluster_info=cluster_info))
+
+        for build_info in worker_builds:
+            build_info.wait_to_schedule()
+
+        thread_pool = ThreadPool(len(worker_builds))
+        # Stream logs for each worker build in real time
+        thread_pool.map(
+            lambda f: f(self.log),
+            [build_info.watch_logs for build_info in worker_builds])
+
+        # Ensure that all builds actually finished
+        for build_info in worker_builds:
+            build_info.wait_to_finish()
+
+        failed_builds = []
+        annotations = {'worker-builds': {}}
+        fail_reasons = {}
+
+        for build_info in worker_builds:
+            annotations['worker-builds'][build_info.platform] = build_info.get_annotations()
+
+            if build_info.build.is_failed():
+                self.log.error('worker build %s failed on %s',
+                               build_info.name, build_info.cluster.name)
+                fail_reasons[build_info.platform] = build_info.get_fail_reason()
+
+        if fail_reasons:
+            return BuildResult(fail_reason=json.dumps(fail_reasons))
```

Comment:
When we create the BuildResult we should also provide `logs` - this is intended to end up in `build.log` in the Koji build.